### ISSUE TITLE
Password generator to main section

### ISSUE THREAD

**Issue by TahirJalilov** [Kunzisoft/KeePassDX#1234]
It will be very useful to have password generator section in the main menu.

Currently it's not possible to use KeePassDX if I want to register somewhere, because there's no password generator.

**Comment by J-Jamet** [comment]
> Currently it's not possible to use KeePassDX if I want to register somewhere, because there's no password generator.

The password generator is available when creating an entry.

> It will be very useful to have password generator section in the main menu.

I see a limited interest in generating passwords that will be lost but I plan to make another application just for password generation and create a separate icon on the launcher.

**Comment by TahirJalilov** [comment]
> The password generator is available when creating an entry.

Yes, I know it. But it's not useful when you need to register somewhere. Or you prefer first add entry, generate password, copy it and then use copied password during registration?

> I see a limited interest in generating passwords that will be lost but I plan to make another application just for password generation and create a separate icon on the launcher.

It should not be lost. KeePassDX should have possibility to generate and copy the password. At least will be nice to have password generator in inline mode.
I don't understand why you think that create separate password generator app is better than just add already exist generator to the main menu of the KeePassDX?

Other password managers apps has this option.

**Comment by J-Jamet** [comment]
> Yes, I know it. But it's not useful when you need to register somewhere. Or you prefer first add entry, generate password, copy it and then use copied password during registration?

Copying to the clipboard is very strongly discouraged as explained in this wiki page : https://github.com/Kunzisoft/KeePassDX/wiki/Clipboard
Encouraging this practice would encourage constant use of the clipboard while KeePassDX is a security-oriented application.

> At least will be nice to have password generator in inline mode.

The Autofill workflow already works for entry registration, so for the creation of a password generated from an entry. (If you don't mean autofill, what do you mean by inline mode?)

> I don't understand why you think that create separate password generator app is better than just add already exist generator to the main menu of the KeePassDX?

I didn't say it was better to create a new app than add an icon to the history screen of KeePassDX.
I just said I'll refactor the password generator code and add an icon on the Android launcher.
Might as well have an icon on the launcher rather than on the database history screen if a database is not going to be opened anyway. (and how to access it if a database is bookmarked since the screen is skipped?)

The new password generator application will be more complete and will contain dictionaries for passphrases in several languages. If installed on the device, it will improve the internal password generator of KeePassDX so used independently as you like.

> Other password managers apps has this option.

This is not a reason if functionally it is not suitable. If the neighbor jumps out of the window, do you also do it without thinking? If you describe to me a more secure and faster workflow, we can explore solutions.

**Comment by TahirJalilov** [comment]
Well, let's discuss in other way. I'll list you the steps that I have to do during registration somewhere.
1) Open web page or mobile app where I need to create new account
2) Fill username
3) Open KeePassDX
4) Go to some exist entry
5) Press edit button
6) Press password generator button
7) Copy generated password from the popup window
8) Press discard changes button (to not save new password for to exist entry
9) After that, I can use this generated and strong password to fill it on ihe registration window
10) If registration success, I have to add new entry manually to the database. Paste copied password and username.


Ho do you think, is it correct way? Maybe the problem with my phone, but I don't see any autofill button on password field during registration, and KeePassDX does'n ask me to store in db newly registered entry.

**Comment by J-Jamet** [comment]
From your workflow :
4. Why use an existing entry? Your goal is to create a new entry. So you might as well create one and save it, pretty logical. ;) 
7. As I said it is completely discouraged to copy a password into the clipboard.
8. 9. 10 No interest if you have created an entry.

If your problem is that autofill is not displaying a popup, the form, browser or application may be preventing the autofill service from the device.

In this case, use the Magikeyboard : 
1. From the browser, share to KeePassDX to prefill the URL in the new entry.
2. Fill the new entry, choose your password and other fields.
3. Validate
4. Back to the Magikeyboard
5. Click on each button on the keyboard to fill in the fields on your web form.

For detailed explanations (with screenshots) : https://github.com/Kunzisoft/KeePassDX/wiki/Magikeyboard
If the registration autofill is not blocked for your form: https://github.com/Kunzisoft/KeePassDX/wiki/AutoFill

**Comment by TahirJalilov** [comment]
> 4. Why use an existing entry? Your goal is to create a new entry. So you might as well create one and save it, pretty logical. ;)

You mean, better to create a new entry, save it in db before registration? And then delete it if registration will unsuccessful?
I don't think that everyone do it like that. At least when using desktop app, normal flow is register first and then give permission for plugin to add new entry to the db.

**Comment by J-Jamet** [comment]
> You mean, better is to create a new entry, save it in db before registration?

Yes, If you call "registration" the confirmation of the service you create an account for, I don't see any problem with that. 

> And then delete it if registration will unsuccessful?

You give up completely your registration to a service if you have an error that appears in the validation? I don't think it happens that often but in this case yes, just delete the entry, it's not a big deal.

> At least when using desktop app, normal flow is register first and then give permission for plugin to add new entry to the db.

Computers are much more permissive than the Android system. If you have a better technical method with all the constraints of the system, feel free to propose your solution.

**Comment by TahirJalilov** [comment]
Ok @J-Jamet  let me try to change registration flow in that way ))) But anyway I think that it will be nice to add password generator button somewhere in the main page under "+" button)))
![image](https://user-images.githubusercontent.com/43037294/153190721-7b6664a1-a165-4dc3-997e-fc620011335b.png)

**Comment by J-Jamet** [comment]
Hoo, we didn't understand each other on the main page already, I thought you were talking about the database selection page.
In any case, for me, if the generation of the password at this moment is neither linked to an entry nor to a database, there is no reason to put it on this screen.
I think it should be done earlier, either by proposing a "generate free password" button (or something like that) in the autofill and in the keyboard. And if autofill is not available and you don't use the keyboard, I suggest a password generator icon on the launcher (not even a database needs to be open). This way you click on the password generator icon instead of the password manager icon.

**Comment by TahirJalilov** [comment]
Yesss!! Exactly what I suggest ))) sorry for my bad English ))
Thank you for your patience )))

**Comment by J-Jamet** [comment]
I thought about this feature and in fact the interest is very limited to create an independent password generator application.
As you normally don't have to copy passwords in the clipboard, I don't see how to pass the generated passwords in the forms without using the Magikeyboard or the Autofill.

So it's only useful for generating passwords that are temporary, but it doesn't fit your need to register in a service.

**Comment by Aeyk** [comment]
I would also like this feature, or at least generator option for custom fields. I use fields for security questions, and would rather use secret passwords then discoverable answers to the common security questions.

**Comment by Pflanzenritter29** [comment]
Heya @J-Jamet 👋

I am tagging you here because this is the closest feature request that already exists;
Could you make the password generator available upon/within the database creation flow?

If I'm a new user, I would like to have the entire setup process for my database - including creating a randomly generated, strong password - within the app.

KeePassXC and KeePass2Android both have this, so this was the first thing that came to mind, when I first used DX, that felt like a fundamental feature missing, to me..

**Comment by J-Jamet** [comment]
> Could you make the password generator available upon/within the database creation flow?

This is already the case, you need to create an entry and generate the password inside. Apart from that, I'm aware that the popup learning system needs to be redone.

**Comment by Pflanzenritter29** [comment]
> > Could you make the password generator available upon/within the database creation flow?
> 
> This is already the case, you need to create an entry and generate the password inside. Apart from that, I'm aware that the popup learning system needs to be redone.

Ah, no, I'm refering to when you have zero databases even created, so you want to create a database directly in KeePassDX, and if it's possible to make the generator available for generating the master password

**Comment by J-Jamet** [comment]
I think it's dangerous to use a password generator to create texts that can't be saved. People would be encouraged to create random text and not remember their main password. Otherwise, we have to change the incentives and make it clear that the generated password must be absolutely remembered by the user.

**Comment by Pflanzenritter29** [comment]
I guess a pop-up/warning, with a 3-second countdown, before it's clickable, could make this work(?) Or just a setting to enable a power-user option like that?

For people who know what they're doing, it's just tedious, having to create a seperate database just to generate strong master passwords within KeePassDX.
I was really thrown off at first, like I said, because that is a standard feature with XC and 2Android.

**Comment by Pflanzenritter29** [comment]
> I think it's dangerous to use a password generator to create texts that can't be saved. People would be encouraged to create random text and not remember their main password. Otherwise, we have to change the incentives and make it clear that the generated password must be absolutely remembered by the user.

I just thought about this again, and what about just asking to re-type the password. GrapheneOS does this for their device-backup flow for example.

**Comment by ghost** [comment]
This has gotten a bit out of hand. The original request is partly based on a false premise. If you register a new account, use the magic keyboard. If it it's on a website, use the share button in your browser. In any other app, switch directly to the magic keyboard, as explained [years ago](https://github.com/Kunzisoft/KeePassDX/issues/1234#issuecomment-1033601347). This was rejected by a strawman [argument](https://github.com/Kunzisoft/KeePassDX/issues/1234#issuecomment-1033627724).

Providing a standalone password generator within this context would encourage insecure and unsafe practices as explained by the developer. Insecure meaning to use the clipboard for passwords. Unsafe meaning the very likely risk of not having the password stored in the database which is one of the very core features.

KeePass,  KeePassXC are not mobile apps but desktop apps. They don't have a magic keyboard. Passwords in clipboards are a terrible practice especially since clipboards feature a history nowadays and are anything but temporary. Auto-deleting the clipboard as a mitigation is a delusion.

**Comment by Pflanzenritter29** [comment]
@cbiere Ah, no, I am actually talking about something slightly different, but decided to first start discussing it here, because it was the closest already existing pull-request.

I am talking about having a password generator available upon database creation, and followed up with the idea to have the user confirm they have saved/written down that master password somewhere, by asking for it again, before creating the database..

@J-Jamet (I am new to discussing on GitHub, so excuse me if I don't know the basics) Should we move my request to a seperate entry/should I create a seperate one, so there isn't any confusion?

**Comment by J-Jamet** [comment]
Indeed, a new query per request is much better for tracking and processing subjects correctly. Can you open a new issue?

**Comment by Pflanzenritter29** [comment]
@J-Jamet I'll - most likely - do that tomorrow and tag you 👋

**Comment by Pflanzenritter29** [comment]
@J-Jamet I have created an issue: #2177 👋👍
